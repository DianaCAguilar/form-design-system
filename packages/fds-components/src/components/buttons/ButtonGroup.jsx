import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import GroupButton from './GroupButton';

/* Allows you to combine functions */
function combine(...functions) {
  return (...args) =>
    functions
      .filter((func) => typeof func === 'function')
      .forEach((func) => func(...args));
}

const ButtonGroup = ({ activeValue, buttons, onChange, className, ...restBtnGroup }) => {
  const rootClass = cx('btngroup', className);
  return (
    <div className={rootClass} {...restBtnGroup}>
      {buttons.map(({ value, content, icon, onClick, ...restBtn }) => (
        <GroupButton
          key={value}
          isActive={activeValue === value}
          onClick={combine(() => onChange(value), onClick)}
          label={content || value}
          Icon={icon}
          {...restBtn}
        />
      ))}
    </div>
  );
};

ButtonGroup.propTypes = {
  activeValue: PropTypes.string,
  /** Each object in array accepts value, content, onClick, and icon, and also will pass through any extra props */
  buttons: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string,
      content: PropTypes.string,
      icon: PropTypes.func,
    })
  ).isRequired,
  className: PropTypes.string,
  onChange: PropTypes.func,
};

export default ButtonGroup;
