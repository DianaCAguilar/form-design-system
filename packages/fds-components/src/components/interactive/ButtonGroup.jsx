import React from 'react';
import PropTypes from 'prop-types';
import GroupButton from './GroupButton';
import combine from '../../util/combine';

const ButtonGroup = ({ buttons, onChange, ...restBtnGroup }) => (
  <div {...restBtnGroup} className="btngroup">
    {buttons.map(({ value, content, key, Icon, icon, onClick, label, ...restBtn }) => (
      <GroupButton
        key={key || label || value}
        onClick={onChange ? combine(() => onChange(label || value), onClick) : onClick}
        label={label || content || value}
        Icon={Icon || icon}
        {...restBtn}
      />
    ))}
  </div>
);

ButtonGroup.propTypes = {
  /** Each object in array accepts value, content, onClick, and icon, and also will pass through any extra props */
  buttons: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * This allows for overriding of the Chip root element
       * (Meant to accomodate `<Link />`)
       */
      as: PropTypes.func,
      /**
       * Use if labels are not unique
       */
      key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      isActive: PropTypes.bool,
      /**
       * `content` & `value` props are deprecated
       */
      label: PropTypes.string,
      /**
       * deprecated (use `Icon` instead)
       */
      icon: PropTypes.func,
      Icon: PropTypes.func,
    })
  ).isRequired,
  onChange: PropTypes.func,
};

export default ButtonGroup;
