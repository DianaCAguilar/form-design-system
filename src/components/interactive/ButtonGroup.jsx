import React from 'react';
import PropTypes from 'prop-types';
import GroupButton from './GroupButton';
import combine from '../../util/combine';

const ButtonGroup = ({ buttons, onChange, ...restBtnGroup }) => (
  <div {...restBtnGroup} className="btngroup">
    {buttons.map(({ value, content, key, Icon, onClick, label, ...restBtn }) => (
      <GroupButton
        key={key || label}
        onClick={onChange ? combine(() => onChange(label || value), onClick) : onClick}
        label={label}
        Icon={Icon}
        {...restBtn}
      />
    ))}
  </div>
);

ButtonGroup.propTypes = {
  /** Each object in array accepts value, content, onClick, and icon, and also will pass through any extra props */
  buttons: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * This allows for overriding of the Chip root element
       * (Meant to accomodate `<Link />`)
       */
      as: PropTypes.func,
      /**
       * Use if labels are not unique
       */
      key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      isActive: PropTypes.bool,
      label: PropTypes.string,
      Icon: PropTypes.func,
    })
  ).isRequired,
  onChange: PropTypes.func,
};

export default ButtonGroup;
