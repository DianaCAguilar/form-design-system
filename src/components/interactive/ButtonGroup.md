# ButtonGroup

Displays a segmented button list (rounded).
