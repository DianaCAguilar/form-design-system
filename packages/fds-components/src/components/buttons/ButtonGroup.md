```js
import { useState } from 'react';

import ButtonGroup  from './ButtonGroup';

import FeedIcon from 'fds-icons/lib/react/FeedIcon';
import TableIcon from 'fds-icons/lib/react/TableIcon';
import MarketMapIcon from 'fds-icons/lib/react/MarketMapIcon';
import StarEmptyIcon from 'fds-icons/lib/react/StarEmptyIcon';

const custom = () => {
  alert('all unknown props will pass through')
}

const buttons = [
  { value: 'Feed', icon: FeedIcon},
  { value: 'Table', icon: TableIcon},
  { value: 'Map', icon: MarketMapIcon},
  { value: 'Suggested', icon: StarEmptyIcon, onClick: custom },
];

const [firstButton] = buttons;

const [activeValue, setActiveValue] = useState(firstButton.value);

<ButtonGroup buttons={buttons} activeValue={activeValue} onChange={setActiveValue} />
```