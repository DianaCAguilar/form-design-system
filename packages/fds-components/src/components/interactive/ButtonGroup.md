Hardcoded example without any local state

```js
import ButtonGroup  from './ButtonGroup';

import FeedIcon from 'fds-icons/lib/react/FeedIcon';
import TableIcon from 'fds-icons/lib/react/TableIcon';
import MarketMapIcon from 'fds-icons/lib/react/MarketMapIcon';
import StarEmptyIcon from 'fds-icons/lib/react/StarEmptyIcon';

// Yes, this even works for Link components
const Link = ({ to, children, ...rest }) => (
  <a href={to} {...rest}>{children}</a>
)

const buttons = [
  { label: 'Feed', Icon: FeedIcon, isActive: true, isActive: true, href: "#" },
  { label: 'Table', Icon: TableIcon, disabled: true, isActive: false,  },
  { label: 'Map', Icon: MarketMapIcon, isActive: false, as: Link, to: '#example' },
  { label: 'Suggested', Icon: StarEmptyIcon, onClick: () => console.log('more custom logic here'), isActive: false },
];

<ButtonGroup buttons={buttons} onChange={() => console.log('custom logic here')} />
```

Example wired up to local state

```js
import { useState } from 'react';

import ButtonGroup  from './ButtonGroup';

import FeedIcon from 'fds-icons/lib/react/FeedIcon';
import TableIcon from 'fds-icons/lib/react/TableIcon';
import MarketMapIcon from 'fds-icons/lib/react/MarketMapIcon';
import StarEmptyIcon from 'fds-icons/lib/react/StarEmptyIcon';

const [activeButton, setActiveButton] = useState('Feed');

const buttons = [
  { label: 'Feed', Icon: FeedIcon, isActive: activeButton === 'Feed' },
  { label: 'Table', Icon: TableIcon, disabled: true },
  { label: 'Map', Icon: MarketMapIcon, isActive: activeButton === 'Map' },
  { label: 'Suggested', Icon: StarEmptyIcon, onClick: () => console.log('more custom logic here'), isActive: activeButton === 'Suggested' },
];

<ButtonGroup buttons={buttons} onChange={setActiveButton} />
```