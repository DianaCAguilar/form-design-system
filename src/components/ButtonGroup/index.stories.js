import React from 'react';
import Tooltip from 'components/Tooltip';
import FeedIcon from 'icons/react/FeedIcon';
import State from 'util/storybook-docs/State';
import ButtonGroup from '.';

const Template = (args) => <ButtonGroup {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  buttons: [
    { label: 'Lorem', disabled: true },
    { label: 'Ipsum' },
    { label: 'Dolor', isActive: true },
  ],
};

export const ActiveStates = () => (
  <State
    initialValue={1}
    render={(value, setValue) => (
      <ButtonGroup
        buttons={['Lorem', 'Sit', 'Ipsum'].map((label, idx) => ({
          label,
          onClick: () => setValue(idx),
          isActive: value === idx,
        }))}
      />
    )}
  />
);

export const Wrapper = Template.bind({});
Wrapper.args = {
  buttons: [
    {
      label: 'Tooltip',
      // eslint-disable-next-line
      Wrapper: (props) => (
        // eslint-disable-next-line
        <Tooltip trigger={props.children} message="hey" maxWidth={240} />
      ),
    },
  ],
};

Wrapper.parameters = {
  docs: {
    description: {
      story: 'Allows you to render a component around any of the individual buttons',
    },
  },
};

export const Icons = Template.bind({});
Icons.args = {
  buttons: [
    { label: 'Lorem', Icon: FeedIcon },
    { label: 'Sit', Icon: FeedIcon },
  ],
};

Icons.parameters = {
  docs: {
    description: {
      story: 'Displays on the left',
    },
  },
};

export default {
  component: ButtonGroup,
  title: 'components/ButtonGroup',
};
