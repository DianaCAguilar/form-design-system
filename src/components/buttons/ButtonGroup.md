Hardcoded example without any local state

```js
import { useState } from 'react';

import ButtonGroup  from './ButtonGroup';

import FeedIcon from '../../../lib/icons/react/FeedIcon';
import TableIcon from '../../../lib/icons/react/TableIcon';
import MarketMapIcon from '../../../lib/icons/react/MarketMapIcon';
import StarEmptyIcon from '../../../lib/icons/react/StarEmptyIcon';

const buttons = [
  { value: 'Feed', icon: FeedIcon, isActive: true, isActive: true },
  { value: 'Table', icon: TableIcon, disabled: true, isActive: false },
  { value: 'Map', icon: MarketMapIcon, isActive: false },
  { value: 'Suggested', icon: StarEmptyIcon, onClick: () => console.log('more custom logic here'), isActive: false },
];

<ButtonGroup buttons={buttons} onChange={() => console.log('custom logic here')} />
```

Example wired up to local state

```js
import { useState } from 'react';

import ButtonGroup  from './ButtonGroup';

import FeedIcon from '../../../lib/icons/react/FeedIcon';
import TableIcon from '../../../lib/icons/react/TableIcon';
import MarketMapIcon from '../../../lib/icons/react/MarketMapIcon';
import StarEmptyIcon from '../../../lib/icons/react/StarEmptyIcon';

const [activeValue, setActiveValue] = useState('Feed');

const buttons = [
  { value: 'Feed', icon: FeedIcon, isActive: activeValue === 'Feed' },
  { value: 'Table', icon: TableIcon, disabled: true },
  { value: 'Map', icon: MarketMapIcon, isActive: activeValue === 'Map' },
  { value: 'Suggested', icon: StarEmptyIcon, onClick: () => console.log('more custom logic here'), isActive: activeValue === 'Suggested' },
];

<ButtonGroup buttons={buttons} onChange={setActiveValue} />
```
