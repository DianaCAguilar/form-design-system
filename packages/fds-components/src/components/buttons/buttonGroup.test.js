import React from 'react';
import { shallow } from 'enzyme';

import ButtonGroup from './ButtonGroup';

const renderComponent = (props) => shallow(<ButtonGroup {...props} />);

const Icon = () => (
  <span>ðŸŒ­</span>
)

const buttons = [
  { value: 'Feed', icon: Icon},
  { value: 'Table', icon: Icon, onClick: () => console.log('additional console')},
];

describe('ButtonGroup component', () => {

  it('matches snapshot (set all props)', () => {
    const component = renderComponent({
      activeValue: 'Feed',
      buttons,
      onChange: () => console.log('changed')
    });
    expect(component).toMatchSnapshot();
  });
});
