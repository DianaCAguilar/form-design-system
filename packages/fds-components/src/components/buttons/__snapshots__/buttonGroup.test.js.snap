// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup component matches snapshot (default props) 1`] = `
<div
  className="btngroup"
>
  <GroupButton
    Icon={[Function]}
    isActive={true}
    key="Feed"
    label="Feed"
    onClick={[Function]}
  />
  <GroupButton
    Icon={[Function]}
    disabled={true}
    key="Table"
    label="Table"
    onClick={[Function]}
  />
</div>
`;

exports[`ButtonGroup component matches snapshot (set all props) 1`] = `
<div
  className="btngroup foo"
>
  <GroupButton
    Icon={[Function]}
    isActive={true}
    key="Feed"
    label="Feed"
    onClick={[Function]}
  />
  <GroupButton
    Icon={[Function]}
    disabled={true}
    key="Table"
    label="Table"
    onClick={[Function]}
  />
</div>
`;
